module/scheduler_components/deterministic_lock_manager.cpp

```diff
@@ -196,6 +196,92 @@ DeterministicLockManager::ReleaseLocks(const Transaction& txn) {
   return ready_txns;
 }
 
+VerifyMasterResult
+DeterministicLockManager::VerifyMaster(const Transaction& txn) {
+  auto keys = ExtractKeys(txn);
+  if (keys.empty()) {
+    // None of the keys in this txn is in this partition
+    return VerifyMasterResult::Valid;
+  }
+
+  auto txn_master_metadata = txn.internal().master_metadata();
+  if (txn_master_metadata.empty()) { // allow this to not be set for testing
+    LOG(WARNING) << "Master metadata empty: txn id " << txn.internal().id();
+    return VerifyMasterResult::Valid;
+  }
+
+  // keeps track of counters that will need to be waited on
+  std::unordered_map<Key, int32_t> ahead_counters;
+  for (auto pair : keys) {
+    auto key = pair.first;
+
+    Record record;
+    bool found = storage_->Read(key, record);
+    if (found) {
+      CHECK(txn_master_metadata.contains(key))
+              << "Master metadata for key \"" << key << "\" is missing";
+
+      auto txn_metadata = txn_master_metadata.at(key);
+      auto stored_metadata = record.metadata;
+
+      if (txn_metadata.counter() > stored_metadata.counter) {
+        ahead_counters[key] += txn_metadata.counter() - stored_metadata.counter;
+      } else if (txn_metadata.counter() < stored_metadata.counter) {
+        return VerifyMasterResult::Aborted;
+      } else {
+        CHECK(txn_metadata.master() == stored_metadata.master)
+                << "Metadata is invalid, different masters with the same counter";
+      }
+    } else { // key not in storage
+      // TODO: verify if counter is 0?
+    }
+  }
+  if (!ahead_counters.empty()){
+    auto txn_id = txn.internal().id();
+    num_remasters_waited_[txn_id] = ahead_counters;
```

Comment:
I think it would work if only the main txn gets sent here by the scheduler? Assuming we take the approach where the entire txn is aborted if counters are behind

Reply:
Yeah, that'd work. Since you mentioned all LO txns will be removed from these queues when an MH txn is aborted so I thought you used them instead. In any case, there need a way to ignore the LO txns that come after the main txn is aborted.

Reply:
I think I would opt to do this in the scheduler- even after abort, keep the txn around until all of it had arrived then delete